File: android_app/app/src/main/java/com/epam/kborodin/bored/data/ActionRepositoryImpl.kt
package com.epam.kborodin.bored.data

import com.epam.kborodin.bored.data.local.dao.ActionDao
import com.epam.kborodin.bored.data.remote.api.BoredApi
import com.epam.kborodin.bored.domain.model.Action
import com.epam.kborodin.bored.domain.model.converter.toDomain
import com.epam.kborodin.bored.domain.model.converter.toEntity
import com.epam.kborodin.bored.domain.repository.ActionRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import javax.inject.Inject

class ActionRepositoryImpl @Inject constructor(
    private val remote: BoredApi,
    private val local: ActionDao
) : ActionRepository {

    override fun getAction(): Flow<Action?> = flow {
        emit(local.getAction()?.toDomain())
        val response = remote.fetchAction()
        local.update(response.toEntity())
        emit(local.getAction()?.toDomain())
    }
}

File: flutter_app/lib/data/repository_impl/action_repository_impl.dart
import 'package:flutter_app/app/locator.dart';
import 'package:flutter_app/data/local/bored_database.dart';
import 'package:flutter_app/data/local/entity/action_entity.dart';
import 'package:flutter_app/data/remote/bored_api.dart';
import 'package:flutter_app/domain/model/action.dart';
import 'package:flutter_app/domain/repository/action_repository.dart';
import 'package:injectable/injectable.dart';

@LazySingleton(as: ActionRepository)
class ActionRepositoryImpl implements ActionRepository {
  final BoredApi _remote = locator<BoredApi>();
  final BoredDatabase _local = locator<BoredDatabase>();

  @override
  Stream<Action> getAction() async* {
    var local = await _local.actionDao.getAction();
    if (local != null) {
      yield Action.fromEntity(local);
    }
    final response = await _remote.fetchAction();
    _local.actionDao.update(
      ActionEntity.fromResponse(response),
    );
    local = await _local.actionDao.getAction();
    if (local != null) {
      yield Action.fromEntity(local);
    }
  }
}
